### ISSUE TITLE
api: /1.0/metrics: lxd_cpu_seconds_total mistake

### ISSUE THREAD

**Issue by exsver** [lxc/lxd#10820]
# Required information

 * Distribution: Ubuntu
 * Distribution version: Ubuntu 22.04.1 LTS
 * The output of "lxc info":
<pre>
# lxc info
config: {}
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_stop_priority
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
- resources
- kernel_limits
- storage_api_volume_rename
- macaroon_authentication
- network_sriov
- console
- restrict_devlxd
- migration_pre_copy
- infiniband
- maas_network
- devlxd_events
- proxy
- network_dhcp_gateway
- file_get_symlink
- network_leases
- unix_device_hotplug
- storage_api_local_volume_handling
- operation_description
- clustering
- event_lifecycle
- storage_api_remote_volume_handling
- nvidia_runtime
- container_mount_propagation
- container_backup
- devlxd_images
- container_local_cross_pool_handling
- proxy_unix
- proxy_udp
- clustering_join
- proxy_tcp_udp_multi_port_handling
- network_state
- proxy_unix_dac_properties
- container_protection_delete
- unix_priv_drop
- pprof_http
- proxy_haproxy_protocol
- network_hwaddr
- proxy_nat
- network_nat_order
- container_full
- candid_authentication
- backup_compression
- candid_config
- nvidia_runtime_config
- storage_api_volume_snapshots
- storage_unmapped
- projects
- candid_config_key
- network_vxlan_ttl
- container_incremental_copy
- usb_optional_vendorid
- snapshot_scheduling
- snapshot_schedule_aliases
- container_copy_project
- clustering_server_address
- clustering_image_replication
- container_protection_shift
- snapshot_expiry
- container_backup_override_pool
- snapshot_expiry_creation
- network_leases_location
- resources_cpu_socket
- resources_gpu
- resources_numa
- kernel_features
- id_map_current
- event_location
- storage_api_remote_volume_snapshots
- network_nat_address
- container_nic_routes
- rbac
- cluster_internal_copy
- seccomp_notify
- lxc_features
- container_nic_ipvlan
- network_vlan_sriov
- storage_cephfs
- container_nic_ipfilter
- resources_v2
- container_exec_user_group_cwd
- container_syscall_intercept
- container_disk_shift
- storage_shifted
- resources_infiniband
- daemon_storage
- instances
- image_types
- resources_disk_sata
- clustering_roles
- images_expiry
- resources_network_firmware
- backup_compression_algorithm
- ceph_data_pool_name
- container_syscall_intercept_mount
- compression_squashfs
- container_raw_mount
- container_nic_routed
- container_syscall_intercept_mount_fuse
- container_disk_ceph
- virtual-machines
- image_profiles
- clustering_architecture
- resources_disk_id
- storage_lvm_stripes
- vm_boot_priority
- unix_hotplug_devices
- api_filtering
- instance_nic_network
- clustering_sizing
- firewall_driver
- projects_limits
- container_syscall_intercept_hugetlbfs
- limits_hugepages
- container_nic_routed_gateway
- projects_restrictions
- custom_volume_snapshot_expiry
- volume_snapshot_scheduling
- trust_ca_certificates
- snapshot_disk_usage
- clustering_edit_roles
- container_nic_routed_host_address
- container_nic_ipvlan_gateway
- resources_usb_pci
- resources_cpu_threads_numa
- resources_cpu_core_die
- api_os
- container_nic_routed_host_table
- container_nic_ipvlan_host_table
- container_nic_ipvlan_mode
- resources_system
- images_push_relay
- network_dns_search
- container_nic_routed_limits
- instance_nic_bridged_vlan
- network_state_bond_bridge
- usedby_consistency
- custom_block_volumes
- clustering_failure_domains
- resources_gpu_mdev
- console_vga_type
- projects_limits_disk
- network_type_macvlan
- network_type_sriov
- container_syscall_intercept_bpf_devices
- network_type_ovn
- projects_networks
- projects_networks_restricted_uplinks
- custom_volume_backup
- backup_override_name
- storage_rsync_compression
- network_type_physical
- network_ovn_external_subnets
- network_ovn_nat
- network_ovn_external_routes_remove
- tpm_device_type
- storage_zfs_clone_copy_rebase
- gpu_mdev
- resources_pci_iommu
- resources_network_usb
- resources_disk_address
- network_physical_ovn_ingress_mode
- network_ovn_dhcp
- network_physical_routes_anycast
- projects_limits_instances
- network_state_vlan
- instance_nic_bridged_port_isolation
- instance_bulk_state_change
- network_gvrp
- instance_pool_move
- gpu_sriov
- pci_device_type
- storage_volume_state
- network_acl
- migration_stateful
- disk_state_quota
- storage_ceph_features
- projects_compression
- projects_images_remote_cache_expiry
- certificate_project
- network_ovn_acl
- projects_images_auto_update
- projects_restricted_cluster_target
- images_default_architecture
- network_ovn_acl_defaults
- gpu_mig
- project_usage
- network_bridge_acl
- warnings
- projects_restricted_backups_and_snapshots
- clustering_join_token
- clustering_description
- server_trusted_proxy
- clustering_update_cert
- storage_api_project
- server_instance_driver_operational
- server_supported_storage_drivers
- event_lifecycle_requestor_address
- resources_gpu_usb
- clustering_evacuation
- network_ovn_nat_address
- network_bgp
- network_forward
- custom_volume_refresh
- network_counters_errors_dropped
- metrics
- image_source_project
- clustering_config
- network_peer
- linux_sysctl
- network_dns
- ovn_nic_acceleration
- certificate_self_renewal
- instance_project_move
- storage_volume_project_move
- cloud_init
- network_dns_nat
- database_leader
- instance_all_projects
- clustering_groups
- ceph_rbd_du
- instance_get_full
- qemu_metrics
- gpu_mig_uuid
- event_project
- clustering_evacuation_live
- instance_allow_inconsistent_copy
- network_state_ovn
- storage_volume_api_filtering
- image_restrictions
- storage_zfs_export
- network_dns_records
- storage_zfs_reserve_space
- network_acl_log
- storage_zfs_blocksize
- metrics_cpu_seconds
- instance_snapshot_never
- certificate_token
- instance_nic_routed_neighbor_probe
- event_hub
- agent_nic_config
- projects_restricted_intercept
- metrics_authentication
- images_target_project
- cluster_migration_inconsistent_copy
- cluster_ovn_chassis
- container_syscall_intercept_sched_setscheduler
- storage_lvm_thinpool_metadata_size
- storage_volume_state_total
- instance_file_head
- instances_nic_host_name
- image_copy_profile
- container_syscall_intercept_sysinfo
- clustering_evacuation_mode
- resources_pci_vpd
- qemu_raw_conf
- storage_cephfs_fscache
- network_load_balancer
- vsock_api
api_status: stable
api_version: "1.0"
auth: trusted
public: false
auth_methods:
- tls
environment:
  addresses: []
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
...
    -----END CERTIFICATE-----
  certificate_fingerprint: 69943e08e85360dfc3ab03c9f37946fe5fbd2d4986c6547d247f6425b38b59a1
  driver: lxc | qemu
  driver_version: 5.0.1 | 7.0.0
  firewall: nftables
  kernel: Linux
  kernel_architecture: x86_64
  kernel_features:
    idmapped_mounts: "true"
    netnsid_getifaddrs: "true"
    seccomp_listener: "true"
    seccomp_listener_continue: "true"
    shiftfs: "false"
    uevent_injection: "true"
    unpriv_fscaps: "true"
  kernel_version: 5.15.0-46-generic
  lxc_features:
    cgroup2: "true"
    core_scheduling: "true"
    devpts_fd: "true"
    idmapped_mounts_v2: "true"
    mount_injection_file: "true"
    network_gateway_device_route: "true"
    network_ipvlan: "true"
    network_l2proxy: "true"
    network_phys_macvlan_mtu: "true"
    network_veth_router: "true"
    pidfd: "true"
    seccomp_allow_deny_syntax: "true"
    seccomp_notify: "true"
    seccomp_proxy_send_notify_fd: "true"
  os_name: Ubuntu
  os_version: "22.04"
  project: default
  server: lxd
  server_clustered: false
  server_event_mode: full-mesh
  server_name: lxd
  server_pid: 1042
  server_version: "5.4"
  storage: dir
  storage_version: "1"
  storage_supported_drivers:
  - name: cephfs
    version: 15.2.16
    remote: true
  - name: dir
    version: "1"
    remote: false
  - name: lvm
    version: 2.03.07(2) (2019-11-30) / 1.02.167 (2019-11-30) / 4.45.0
    remote: false
  - name: zfs
    version: 2.1.4-0ubuntu0.1
    remote: false
  - name: ceph
    version: 15.2.16
    remote: true
  - name: btrfs
    version: 5.4.1
    remote: false
</pre>

# Issue description

lxc query -X GET --wait /1.0/metrics reports metric "lxd_cpu_seconds_total" in seconds, but help says: "The total number of CPU seconds used in milliseconds."

<pre>
# HELP lxd_cpu_seconds_total The total number of CPU seconds used in milliseconds.
# TYPE lxd_cpu_seconds_total counter
lxd_cpu_seconds_total{cpu="0",mode="system",name="ct1003",project="default",type="container"} 29
lxd_cpu_seconds_total{cpu="0",mode="user",name="ct1003",project="default",type="container"} 6
...
</pre>
So... error in help here: https://github.com/lxc/lxd/blob/ca3e42ea5e119a1ad8213e133627d334f6e12562/lxd/metrics/types.go#L138
or error in calculation here: https://github.com/lxc/lxd/blob/ca3e42ea5e119a1ad8213e133627d334f6e12562/lxd/instance/drivers/driver_lxc.go#L6921-L6922

**Comment by tomponline** [comment]
@simondeziel any thoughts or preference on this?

**Comment by simondeziel** [comment]
We want to represent seconds for consistency with the rest of the ecosystem in general (see https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#units-and-base-units and https://www.robustperception.io/who-wants-seconds/).

I checked the code (https://github.com/lxc/lxd/blob/ca3e42ea5e119a1ad8213e133627d334f6e12562/lxd/cgroup/abstraction.go#L261-L344) and the 10^9 division we are doing gives us seconds *if I followed properly*.

In the v1 case, https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v1/cpuacct.html mentions that nanoseconds are used for `cpuacct.usage`. We use `cpuacct.usage_all` but they *seem* to use the same unit (nanoseconds).

In the v2 case, https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#cpu-interface-files says microseconds are used everywhere and the fields we are using even have `_usec` in their name. https://github.com/lxc/lxd/blob/ca3e42ea5e119a1ad8213e133627d334f6e12562/lxd/cgroup/abstraction.go#L327 converts those microseconds to nanoseconds.

tl;dr: we are dealing with seconds so we should fix the `HELP` text.

To verify all this, here's the `top` usage sorted by CPU time of one of my container:

```
root@puppet:~# top

top - 13:10:29 up 9 days, 23:20,  0 users,  load average: 0.00, 0.00, 0.00
Tasks:  67 total,   1 running,  66 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0.0 us,  0.0 sy,  0.0 ni, 99.7 id,  0.2 wa,  0.0 hi,  0.2 si,  0.0 st
MiB Mem :   2560.0 total,    653.6 free,   1684.6 used,    221.8 buff/cache
MiB Swap:      0.0 total,      0.0 free,      0.0 used.    875.4 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                                                                                                                                                                                      
    204 puppet    20   0 3265492   1.1g  28656 S   0.0  43.1 159:16.09 java                                                                                                                                                                                                         
    198 puppetdb  20   0 2870628 486576  23756 S   0.3  18.6  20:04.12 java                                                                                                                                                                                                         
      1 root      20   0  170228   8708   5244 S   0.0   0.3   0:50.66 systemd                                                                                                                                                                                                      
 110210 postgres  20   0  215692  21600  19552 S   0.0   0.8   0:05.36 postgres                                                                                                                                                                                                     
 110216 postgres  20   0   70192   4460   2284 S   0.0   0.2   0:02.86 postgres                                                                                                                                                                                                     
 110215 postgres  20   0  216364   6112   3772 S   0.0   0.2   0:01.43 postgres                                                                                                                                                                                                     
    126 systemd+  20   0   19068   5020   4104 S   0.0   0.2   0:01.26 systemd-network                                                                                                                                                                                              
 110214 postgres  20   0  215692   8808   6740 S   0.0   0.3   0:01.22 postgres                                                                                                                                                                                                     
 110212 postgres  20   0  215816  49560  47416 S   0.0   1.9   0:00.95 postgres                                                                                                                                                                                                     
    127 systemd+  20   0   24608   9240   5128 S   0.0   0.4   0:00.94 systemd-resolve                                                                                                                                                                                              
     59 root      20   0   35736   6796   5860 S   0.0   0.3   0:00.77 systemd-journal                                                                                                                                                                                              
 110213 postgres  20   0  215692   7260   5188 S   0.0   0.3   0:00.61 postgres                                                                                                                                                                                                     
    118 syslog    20   0  154688   2584   2068 S   0.0   0.1   0:00.23 rsyslogd                                                                                                                                                                                                     
 121372 postgres  20   0  218712  19408  15092 S   0.0   0.7   0:00.07 postgres                                                                                                                                                                                                     
 121369 postgres  20   0  218168  16868  13280 S   0.0   0.6   0:00.05 postgres                                                                                                                                                                                                     
 110217 postgres  20   0  216120   5052   2796 S   0.0   0.2   0:00.03 postgres                                                                                                                                                                                                     
 121668 root      20   0    9220   2716   2128 R   0.0   0.1   0:00.03 top
```

That's roughly ~180 minutes of CPU which amounts to ~10800 seconds of `user` time. This is roughly what the metrics API gives me:

```
$ lxc query -X GET --wait xeon:/1.0/metrics | grep seconds_total | grep puppet
lxd_cpu_seconds_total{cpu="0",mode="system",name="puppet",project="default",type="container"} 1231
lxd_cpu_seconds_total{cpu="0",mode="user",name="puppet",project="default",type="container"} 10903
```

@tomponline my only question/concern is why are those always integer?

**Comment by tomponline** [comment]
@simondeziel you want them as floats?

**Comment by simondeziel** [comment]
@tomponline yes, float64 is what is stored by Prometheus (https://prometheus.io/docs/concepts/data_model/#samples). It can deal with int but that's leaving precision on the table.

**Comment by tomponline** [comment]
@exsver is this something you would be willing to look at?

**Comment by exsver** [comment]
For cpu_time in seconds, it would be better to use floats.

`float64(stats.System / 1000000000)` -> `float64(stats.System) / float64(1000000000)`
`float64(stats.User / 1000000000)` -> `float64(stats.User) / float64(1000000000)`
